### PR TITLE
MueLu performance tweaks

### PR DESCRIPTION
**Pull request by CamelliaDPG**
@trilinos/muelu @tawiesn @jhux2.  Just an optimization to avoid calls to Teuchos::TypeNameTraits<T>::name(), whose result is only used when exceptions are thrown, when said exceptions are not thrown.  These were showing up as measurable costs in profiling some benchmarks.

### CODE REVIEW

File: packages/muelu/src/Utils/MueLu_Monitor.hpp [hunk 1]
```diff
@@ -115,6 +115,21 @@ namespace MueLu {
...
+     @param[in] objectDescription     Nominally, the result of object.description(), though this is not enforced
+     @param[in] objectShortClassName  Nominally, the result of object.ShortClassName(), though this is not enforced
+     @param[in] msgLevel              Governs whether information should be printed.
+     @param[in] timerLevel            Governs whether timing information should be *gathered*.  Setting this to NoTimeReport prevents the creation of timers.
+     */
+    Monitor(const BaseClass& object, const std::string & msg, const std::string &objectDescription, const std::string &objectShortClassName,
+            MsgType msgLevel = Runtime0, MsgType timerLevel = Timings0)
+      : printMonitor_(object, msg + " (" + objectDescription + ")", msgLevel),
+        timerMonitor_(object, objectShortClassName + ": " + msg + " (total)",    timerLevel)
+    { }
```
Comment (aprokop) [inline packages/muelu/src/Utils/MueLu_Monitor.hpp]: This code does not belong to this PR.

File: packages/muelu/src/MueCentral/MueLu_VariableContainer.hpp [hunk 2]
```diff
@@ -243,31 +251,45 @@ namespace MueLu {
       typedef Teuchos::RCP<Operator> TO;
       typedef Teuchos::RCP<Matrix>   TM;
 
-      const std::string typeTOName = Teuchos::TypeNameTraits<TO>::name();
-      const std::string typeTMName = Teuchos::TypeNameTraits<TM>::name();
-      TEUCHOS_TEST_FOR_EXCEPTION(data_ == NULL, Teuchos::bad_any_cast,
-                                 "Error, cast to type Data<" << typeTOName << "> failed since the content is NULL");
+      if (data_ == NULL) // NVR added guard to avoid unnecessary call to TypeNameTraits<TO>::name(), which calls Teuchos::demangleName(), which can be expensive
```
Comment (aprokop) [inline packages/muelu/src/MueCentral/MueLu_VariableContainer.hpp]: Can you clean this up a bit further, similar to the the above `get` call? Change this `if` to
```c++
if (data_ == NULL && data_->type() != typeid(TO) && data_->type() != typeid(TM)) {
 // throw appropriate exceptions
}
```
then deal with `TO`/`TM`.

**Comment by aprokop** [comment]
This is a nice catch. I'm curious, could you please elaborate on situations when `name()` takes a significant amount of time? I cannot recall any of such situations.

**Comment by aprokop** [comment]
@tawiesn I disagree with hiding this behind `Debug`. These are important potential runtime problems, and it costs us nothing to enable them in non-debug version. This PR properly addresses the issue.

File: packages/muelu/src/Utils/MueLu_Monitor.hpp (hunk 1 above)
Comment (CamelliaDPG) [inline packages/muelu/src/Utils/MueLu_Monitor.hpp]: I can submit a separate PR for this if you like.  I had intended it to be part of the same, but you're right that this is a distinct matter.

File: packages/muelu/src/MueCentral/MueLu_VariableContainer.hpp (hunk 2 above)
Comment (CamelliaDPG) [inline packages/muelu/src/MueCentral/MueLu_VariableContainer.hpp]: I think you mean `||`, not `&&`.  But even then, this will not do quite the same thing as what I am proposing: in the event that `(data_ != NULL)` and `(data_->type() == typeid(TO))` but `dynamic_cast<Data<TO>*>(data_) != NULL`, there will still be an unnecessary call to `name()` in the argument to the first `TEUCHOS_TEST_FOR_EXCEPTION()`, despite the fact that no exception will be thrown.

**Comment by CamelliaDPG** [comment]
@aprokop The application I'm working with has some benchmarks that end up using MueLu as a preconditioner for CG with some fairly ill-conditioned H(Grad) matrices: between 100 and 300 iterations to converge.  We see significant costs from `name()` in cases where these matrices are relatively small--in one case, the matrix was 1448x1448, with nnz=17344.

File: packages/muelu/src/MueCentral/MueLu_VariableContainer.hpp (hunk 2 above)
Comment (CamelliaDPG) [inline packages/muelu/src/MueCentral/MueLu_VariableContainer.hpp]: @aprokop—just checking on this.  As requested, I reverted the MueLu_Monitor changes, but I'd like to leave the MueLu_VariableContainer changes as they are, for the reasons mentioned above.  Does that sound OK?

**Comment by aprokop** [comment]
Nice. Just squash the commits in this PR, and lets merge it.

**Comment by CamelliaDPG** [comment]
@aprokop, I did a squash-merge from this PR's branch into develop, and am running the check-in script now.  Is that what you intended?  (I'm fairly new to the PR process.)

**Comment by aprokop** [comment]
@CamelliaDPG Yes, that would work too. I'm not 100% certain on the current Trilinos' policy, specifically, whether PRs are used just for code review before running checkin script, or for merging. Either way, I think you did the right thing by squashing the commits.

**Comment by CamelliaDPG** [comment]
The checkin script succeeded; these changes have now been pushed.  Thanks all!
